src/multipart/multipart.rs

```diff
@@ -0,0 +1,469 @@
+    ///
+    /// ```
+    /// reqwest::MultipartField::param("key", "value");
+    /// ```
+    ///
+    /// ```
+    /// let key: String = "key".to_string();
+    /// let value: String = "value".to_string();
+    /// reqwest::MultipartField::param(key, value);
+    /// ```
+    ///
+    pub fn param<T: Into<Cow<'static, str>>, U: AsRef<[u8]> + Send + 'static>(name: T, value: U) -> MultipartField {
+        let value_length = Some(value.as_ref().len() as u64);
+        MultipartField {
+            name: name.into(),
+            value: Box::new(std::io::Cursor::new(value)),
+            value_length: value_length,
+            mime: None,
+            file_name: None,
+        }
+    }
+    /// Adds a generic reader.
+    ///
+    /// ```
+    /// use std::io::empty;
+    /// let reader = empty();
+    /// reqwest::MultipartField::reader("key", reader);
+    /// ```
+    ///
+    pub fn reader<T: Into<Cow<'static, str>>, U: Read + Send + 'static>(name: T, value: U) -> MultipartField {
+        MultipartField {
+            name: name.into(),
+            value: Box::from(value),
+            value_length: None,
+            mime: None,
+            file_name: None,
+        }
+    }
+    /// Makes a file parameter.
+    /// Defaults to mime type application/octet-stream.
+    ///
+    /// ```no_run
+    /// reqwest::MultipartField::file("key", "/path/to/file");
+    /// ```
+    ///
+    /// # Errors
+    /// Errors when the file cannot be opened.
+    pub fn file<T: Into<Cow<'static, str>>, U: AsRef<std::path::Path>>(
+        name: T,
+        path: U,
+    ) -> std::io::Result<MultipartField> {
+        let file_name = path.as_ref().file_name().and_then(|filename| {
+            Some(Cow::from(filename.to_string_lossy().into_owned()))
+        });
+        let file_length = std::fs::metadata(path.as_ref())
+            .ok()
+            .and_then(|metadata| Some(metadata.len() as u64));
+        Ok(MultipartField {
+            name: name.into(),
+            value: Box::new(std::fs::File::open(path.as_ref())?),
+            value_length: file_length,
+            mime: Some(::hyper::mime::APPLICATION_OCTET_STREAM),
+            file_name: file_name,
+        })
+    }
+    /// Sets the mime, builder style.
+    ///
+    /// ```
+    /// use reqwest::mime;
+    /// reqwest::MultipartField::param("key", "value").mime(Some(mime::IMAGE_BMP));
+    /// ```
+    ///
+    pub fn mime(mut self, mime: Option<Mime>) -> MultipartField {
+        self.mime = mime;
+        self
+    }
+    /// Sets the filename, builder style.
+    ///
+    /// ```
+    /// reqwest::MultipartField::param("key", "value").file_name(Some("filename"));
+    /// ```
+    ///
+    /// ```
+    /// let file_name = "file_name".to_string();
+    /// reqwest::MultipartField::param("key", "value").file_name(Some(file_name));
+    /// ```
+    ///
+    pub fn file_name<T: Into<Cow<'static, str>>>(mut self, filename: Option<T>) -> MultipartField {
```

Comment:
A user might add a file via `file` which automatically sets the filename since that is the most common expectation and then want to remove the filename.

We could also remove the option from `file_name` and add another `no_file_name()` function.